artifact/tar/platform/libarchive/wrapper.cpp

```diff
@@ -0,0 +1,142 @@
+// Copyright 2023 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#include <archive.h>
+
+#include <libarchive/wrapper.hpp>
+#include <common/log.hpp>
+
+namespace mender {
+namespace libarchive {
+namespace wrapper {
+
+size_t libarchive_read_buffer_size {4096};
+
+namespace expected = mender::common::expected;
+
+using ExpectedSize = expected::ExpectedSize;
+
+// The reader_callback is invoked whenever the library requires raw bytes from
+// the archive. The read callback reads data into a buffer, sets the const void
+// **buffer argument to point to the available data, and returns a count of the
+// number of bytes available. LibArchive will invoke the read callback again
+// only after it has consumed this data. LibArchive imposes no constraints on
+// the size of the data blocks returned.
+//
+// - On EOF return 0.
+// - On error return -1.
+ssize_t reader_callback(archive *archive, void *in_reader_container, const void **buff) {
+	ReaderContainer *p_reader_container = static_cast<ReaderContainer *>(in_reader_container);
```

Comment:
Wouldn't `reinterpret_cast` be a better fit here?

Reply:
`static_cast` should always be preferred over `reinterpret_cast` if it works, because it is more restrictive. IOW it can catch later mistakes more effectively than `reinterpret_cast` can.

Reply:
But when casting a `void` pointer, `static_cast` does nothing, right? (just trying to learn here)

Reply:
Do you mean runtime checks? It's `dynamic_cast` which does that. `static_cast` never does any runtime checks, but it does check at compile time that the cast is "potentially valid". It allows casts from `void *` because they can be anything, but it does not allow unrelated types such as `Reader *` to `Writer *` or `int *` to `string *`. That is when you need `reinterpret_cast`, which allows practically anything.

Reply:
And in case you wonder: We can not use `dynamic_cast` here, because it cannot cast from `void *`.

There are other reasons not to use it too, mainly that the runtime data it requires grows the binary size quite a bit, and some customers might want to keep that disabled (`-fno-rtti` option IIRC).